pkg/pillar/cmd/zedmanager/zedmanager.go

```diff
@@ -692,8 +692,9 @@ func maybeLatchImageSha(ctx *zedmanagerContext, config types.AppInstanceConfig,
 // return the name with the sha inserted.
 // Note that the sha must be lower case in the OCI reference.
 func maybeInsertSha(name string, sha string) string {
-	if strings.Index(name, "@") != -1 {
-		// Already has a sha
+	if strings.Index(name, "@") != -1 || !strings.Contains(name, ":latest") {
```

Comment:
I am confused. The first part says, "if it already has an `@`, then it should have a hash, so just return it. But the second part says, "if it doesn't include latest then return the name as is"?

So if it includes a different tag, e.g. `:v1.2.3`, or no tag at all, e.g. `docker.io/library/nginx`, then we return it as is? Why would we not insert the sha always, except if there already is a sha?

Reply:
My understanding is if the user specified the tag like ```alpine:3.10``` we should not replace tag with the SHA. If the tag is latest ```alpine:latest``` or no tag/SHA ```docker.io/library/nginx``` we will replace/add the SHA. If my understanding is incorrect then we can close this PR.

Reply:
Let's start it from the other side, since I don't actually know.

What are we trying to accomplish? Where will this be stored?

Reply:
While creating an app instance, we are changing the container name and updating the image SHA in the storage status so that it won't be redownloaded at boot.

Reply:
So `alpine` or `alpine:latest` would have the tag added, but `alpine:3.10` would not? Since tags are mutable, I don't know how that would work.

How would we know if the request to have `alpine:3.10` available means "the latest alpine:3.10 which is different than what is on disk" or "don't change it"? According to what is described above, we would now have a policy of never updating tagless or `latest` (which is the same thing), but always pulling if it has any other tag and the tag changes. That seems counter-intuitive.